Leader Election issue #434

Invictus17 opened the pull request:
Leader Election issue #434
https://github.com/kubernetes-client/python/issues/434

Invictus17 commented:
/assign @mbohlool

Invictus17 commented:
Hi @mbohlool @roycaihw , can you please review my PR?

yliaog commented:
thanks for the PR.

could you please keep the file naming style consistent with the existing code? e.g. LeaderElection is better to switched to leaderelection

Invictus17 commented:
> thanks for the PR.
> 
> could you please keep the file naming style consistent with the existing code? e.g. LeaderElection is better to switched to leaderelection

Sure. I've pushed the changes.

roycaihw commented:
/assign

palnabarun commented:
/assign

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,237 @@
+            # Make sure thread that runs onStoppedLeading callback is stopped
+            if self.OnStoppedLeadingThread:
+                self.OnStoppedLeadingThread.stop()
+                self.OnStoppedLeadingThread.join()
```
`onStoppedLeading` is not used. Could you demonstrate how a user should react to `OnStoppedLeadingThread` in the example?

roycaihw replied:
nvm, I misread the thread logic

roycaihw commented on leaderelection/threadingwithexception.py:
```diff
@@ -0,0 +1,31 @@
+import threading
+import ctypes
```
I'm not very excited if we have to add C compatibility to achieve this

Invictus17 replied:
I can look into a different way to kill threads.

cliffburdick replied:
You should be able to use stop flags and let the client check for it.

Invictus17 replied:
> You should be able to use stop flags and let the client check for it.

Yes, that's similar to the latest update using traces(this commit is outdated now) but I also want to look into https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor as suggested by @yliaog in her review.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,237 @@
+            self.OnStoppedLeadingThread.start()
+
+            # Start to follow
+            self.follow(scheduler)
```
In your implementation, a program can keep switching between being a leader and a follower. However reading the client-go implementation, when the leader fails to renew the lease, it [gives up and doesn't re-join as a follower](https://github.com/kubernetes/kubernetes/blob/f5334fad483856c41ceb1031036a96e96c32fcee/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L278-L279). Should we keep the client-go behavior and let the user decide what to do when a leader loses the lease?

Invictus17 replied:
Sure. I can change the code to have a leader exit and close if failed to update lease.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,237 @@
+    def getLatestLeader(self):
+        getStatus, getResponse = self.electionConfig.lock.Get(name=self.electionConfig.lock.name, namespace=self.electionConfig.lock.namespace)
+        if getStatus:
+            return "leader is " + str(ast.literal_eval(getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey])['holderIdentity'])
```
why do we need `literal_eval`?

Invictus17 replied:
getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey] returns a string. 
literal_eval converts that string to a dictionary so that ['holderIdentity'] can be easily accessed.

cliffburdick replied:
> `getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey]` returns a string.
> `literal_eval `converts that string to a dictionary so that `['holderIdentity']` can be easily accessed.

Would json.loads() work or is it not JSON?

Invictus17 replied:
> > `getResponse.metadata.annotations[self.electionConfig.lock.LeaderElectionRecordAnnotationKey]` returns a string.
> > `literal_eval `converts that string to a dictionary so that `['holderIdentity']` can be easily accessed.
> 
> Would json.loads() work or is it not JSON?

No, it's not JSON.

Invictus17 commented:
Hi @roycaihw , Thanks for your review. Based on your comments I've pushed the suggested changes.

changes:  
1) Coding style - Updated variable and function names.  
2) Threading with traces - This implementation does not depend on ctypes.
3) Leader exits - The leader does not become a follower if it fails to update lease. It now exits after running onStoppedLeading().

yliaog commented on leaderelection/threadingwithtrace.py:
```diff
@@ -0,0 +1,50 @@
+
+
+# citing source: https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/
+class thread_with_trace(threading.Thread):
```
it would be simpler to use https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor than working with threads directly.

it is added in python 3, but i guess it is ok since python 2 is deprecated anyway, you don't have to support python 2.

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,257 @@
+            # Make sure thread that runs onstopped_leading callback is stopped
+            if self.onstopped_leadingthread:
+                self.onstopped_leadingthread.stop()
+                self.onstopped_leadingthread.join()
```
i don't think you need to 'stop', then 'join', i.e., wait for it to stop. 

it's better to implement a simpler leaderelection:
1. right at the start, try to acquire lease to be the leader
2. if not yet a leader, periodically check if it can acquire the lease to be leader
3. if it becomes the leader, call the hook on started_leading
4. try to maintain the leadership by renewing the lease
5. if fail to renew the lease, call the hook on_stopped_leading
6. done with the leaderelection, return

NOTE: the lifecycle of one leaderelection run has two possibilities:
a) it always blocks, waiting to be a leader, but never succeeds
b) it becomes a leader, lead for sometime, then stop leading and return

It is simpler because during one leaderelection run, it can be a leader at most *once*. if become a leader, then somehow lose the leadership, then the whole leaderelection returns. it leave to the caller of the leaderelection to decide what to do after that, the caller may exit completely, or the caller may choose to run another leaderelection.

Invictus17 replied:
Thank you for pointing this out. I should have removed this piece of code when I updated the code for a leader to not follow after losing leadership and exit. Other than that, the leader election logic is identical to what you've described. I'll also make sure that the program let's the user decide whether to exit or run for election again. I'll handle these changes in the next commit.

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,257 @@
+    # Point of entry to Leader election
+    def run(self):
+        # Try to create/ acquire a lock
+        self.try_acquire_or_renew()
```
better to check out the java implementation, https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java

It uses only three threads, or more precisely, three threadpools, but each pool has only a single thread.

yliaog commented on leaderelection/resourcelock/configmaplock.py:
```diff
@@ -0,0 +1,74 @@
+from kubernetes.client.api_client import ApiClient
+
+
+class ConfigMapLock:
```
maybe not now, but unittests need to be added before merging the PR

Invictus17 commented on leaderelection/example.py:
```diff
@@ -0,0 +1,46 @@
+from datetime import timedelta
+
+# Authenticate using config file
+config.load_kube_config(config_file=r"D:\Kubernetes open source - Python client\Go example\kubeconfig.txt")
```
sorry, I forgot to leave that out. Updated now.

Invictus17 commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,257 @@
+        lock_status, lock_response = self.election_config.lock.get(self.election_config.lock.name, self.election_config.lock.namespace)
+
+        # create a default Election record for this candidate
+        self.leader_election_record = self.leaderelector_record(self.election_config)
```
Thank you for pointing it out. I'll update it in the following commit.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,257 @@
+        self.leaderfunction_thread = None
+        
+        # onstopped_leadingthread contains the thread object for onstopped_leading
+        self.onstopped_leadingthread = None
```
on_stopped_leading should not be run in a separate thread

Invictus17 commented:
Hi @roycaihw @yliaog ,
Thank you for your review. I have updated my code based on your comments and discussions. The updates are:
1. Not killing threads and making the logic simpler, similar to the initial Java client.
2. OnStoppedLeading() is not run in a sub-thread.
3. Added a README.
4. Using logging and minor changes pointed out during review.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,243 @@
+
+        # updatelease_schedulerId variable stores the scheduler object id for the update_lease schedule that is repeated
+        # every retry_period seconds by the leader
+        self.updatelease_schedulerId = None
```
is this unused?

Invictus17 replied:
No it is used:
https://github.com/Invictus17/python-base/blob/master/leaderelection/leaderelection.py#L138
https://github.com/Invictus17/python-base/blob/master/leaderelection/leaderelection.py#L182

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,243 @@
+                scheduler_leader = sched.scheduler(time.time, time.sleep)
+                self.lead(scheduler_leader)
+
+    def transition_follower_to_leader(self):
```
we don't need to distinguish `transition_follower_to_leader` and `lead`, same for `follow` and `try_acquire`. These complexity might be useful if we wanted to convert a leader back to a follower

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,243 @@
+
+
+            # keep checking for lease updates every retry_period seconds
+            self.followerlease_checkscheduler = scheduler.enter(int(self.election_config.retry_period), 1, self.check_lease_updates, (scheduler,))
```
previously I thought the scheduler was like the `wait.Until` method that client-go uses, where the scheduling and the real logic are isolated. I found this implementation (scheduling the real logic recursively) hard to follow. What's the benefit against using a simple while loop + sleep like the java client?

Invictus17 replied:
I don't think either has a benefit over the other. Would you suggest using a while + sleep?

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,243 @@
+        # If a lock is already created with that name
+        if lock_status:
+            logging.info("{} is a follower".format(self.election_config.lock.identity))
+            scheduler_follower = sched.scheduler(time.time, time.sleep)
```
there are a total of 4 of these: sched.scheduler(time.time, time.sleep) in the code. would one scheduler be sufficient? would it be better to create just one in the class __init__?

Invictus17 replied:
> there are a total of 4 of these: sched.scheduler(time.time, time.sleep) in the code. would one scheduler be sufficient? would it be better to create just one in the class **init**?

Yes, one would do, I'll update it. @roycaihw had a different opinion about using the sched module. Do you have similar thoughts about it?
https://github.com/kubernetes-client/python-base/pull/206#discussion_r471037636

yliaog replied:
Yes, i agree with his comment. i think it's easier to understand to have the program structure like this:

1. try to acquire lease to be leader in main thread
2. if  get lease, then run lead in a separate thread
3. keep renewing lease in main thread
4. if failed renewing lease, then stop leading in main thread

Invictus17 commented on leaderelection/README.md:
````diff
@@ -0,0 +1,11 @@
+### Command to run
+```python example.py```
+
+Now kill the existing leader. You will see from the terminal outputs that one of the remaining two processes will be elected as the new leader.
````
Thanks for pointing it out. I'll update it.

Invictus17 commented:
Hi @roycaihw @yliaog ,
Thank you for your review. I have updated my code based on your comments and discussions. The updates are:
1. Switched from using sched module to a while + sleep approach.
2. Made the code consistent with the other clients.

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,156 @@
+        
+        # If a lock is already created with that name
+        if lock_status:
+            old_election_record = ast.literal_eval(lock_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey])
```
also ast.literal_eval could throw an exception, better to catch it

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,156 @@
+
+
+            # If This candidate is not the leader and lease duration is yet to finish
+            if str(self.election_config.lock.identity) != self.observed_record['holderIdentity'] and self.observed_time_milliseconds + self.election_config.lease_duration*1000 > int(time.time()*1000):
```
self.observed_time_milliseconds is from this candidate, but it should be from the election record, no?

also self.observed_time_milliseconds is updated to the current time at line 117, so it will always be the current time, and the lease duration will never expire

Invictus17 replied:
> self.observed_time_milliseconds is from this candidate, but it should be from the election record, no?
No, the java and go clients are also keeping a local record of 'observed_time_milliseconds' and not referring to the election record
https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L269-L315

 
> also self.observed_time_milliseconds is updated to the current time at line 117, so it will always be the current time, and the lease duration will never expire
It is only updated if a follower identifies that the lock object has been updated by a leader.
       if old_election_record != self.observed_record:
             # Update self.observed_time_milliseconds & self.observed_record

In case a leader fails to update the lock, self.observed_time_milliseconds will not be updated to current time & after a period of 'leaseDuration' a follower will try to update the lock. 
https://github.com/kubernetes-client/java/blob/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L267-L279

yliaog replied:
ok, the "> int(time.time()*1000):" in the above line is incorrect. time.time() would return the current time, what should be used instead is the time at the start of the function try_acquire_or_renew (https://github.com/kubernetes-client/java/blob/6a2a60ad2ad75fb127874797ea910b90e4a80651/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L236)

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,156 @@
+            # If this candidate is the Leader
+            if str(self.election_config.lock.identity) == self.observed_record['holderIdentity']:
+                # Leader sets acquireTime
+                leader_election_record['acquireTime'] = self.observed_record['acquireTime']
```
self.observed_record['acquireTime'] is not updated, it is still the old_election_record's acquireTime

roycaihw replied:
this line is correct I think. It prevents leader_election_record['acquireTime'] to be "now"

yliaog replied:
ok. i was thinking about setting the 'acquireTime' at the time when the lease is acquired, currently, the 'acquireTime' is set at the line: 104: leader_election_record = self.leaderelector_record(self.election_config)

it probably does not matter though, as there is no wait or loop inside the function try_acquire_or_renew

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,156 @@
+
+            if old_election_record != self.observed_record:
+                self.observed_record = old_election_record
+                self.observed_time_milliseconds = int(time.time()*1000)
```
capture the current time (e.g. `now = time.time()`) at the beginning of this function, and keep using `now`, instead of calling time.time() multiple times.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,156 @@
+
+            # Update object with latest election record
+            lock_response.metadata.annotations[self.election_config.lock.leader_electionrecord_annotationkey] = str(leader_election_record)
+            update_status, update_response = self.election_config.lock.update(self.election_config.lock.name, self.election_config.lock.namespace, lock_response)
```
this is not generalized for different types of locks. Better to make leader_election_record a class or a dict, and pass it directly to the lock methods

roycaihw replied:
the leaderelection logic doesn't need to know about `LeaderElectionRecordAnnotationKey`. It's a implementation detail for the lock

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,182 @@
+
+        return self.update_lock(lock_response, leader_election_record)
+
+    def update_lock(self, lock_response, leader_election_record):
```
`lock_response` can be a field stored in the lock, since it was returned from `lock.get()` and never changed before we passed it back to `lock.update()`. See [client-go](https://github.com/kubernetes/client-go/blob/57f65489c11a4d4a978c93ab369c38f53156888a/tools/leaderelection/resourcelock/configmaplock.go#L41) as an example.

roycaihw commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,182 @@
+            self.observed_time_milliseconds = int(time.time() * 1000)
+            return True
+
+        # A lock exists with that name
+        # Validate lock_record
+        if lock_record is None:
+            # try to update lock with proper annotation and election record
+            return self.update_lock(lock_response, leader_election_record)
+
+        # check for any key, value errors in the record
+        try:
+            old_election_record = ast.literal_eval(lock_record)
+            if (old_election_record['holderIdentity'] == '' or old_election_record['leaseDurationSeconds'] == ''
+                    or old_election_record['acquireTime'] == '' or old_election_record['renewTime'] == ''):
+                # try to update lock with proper annotation and election record
+                return self.update_lock(lock_response, leader_election_record)
+        except:
+            # try to update lock with proper annotation and election record
+            return self.update_lock(lock_response, leader_election_record)
```
I think it's better to assume `lock_record` is a class object, instead of a string here. Each lock type can have its own unmarshalling implementation (e.g. ConfigMap and Endpoints unmarshal their annotations, while Lease unmarshalls its spec). 

This will also allow the short-circuit logic here to be as simple as [client-java](https://github.com/kubernetes-client/java/blob/abea7f4dbf9bf523a4e94349129309da09e317e0/extended/src/main/java/io/kubernetes/client/extended/leaderelection/LeaderElector.java#L259-L264).

roycaihw commented on leaderelection/resourcelock/configmaplock.py:
```diff
@@ -0,0 +1,129 @@
+        """
+        body = client.V1ConfigMap(
+            metadata={"name": name,
+                      "annotations": {self.leader_electionrecord_annotationkey: json.dumps(self.get_lock_dict(election_record))}})
```
`self.get_lock_dict` always updates `self.lock_record`. I think we don't want to update the lock record if the create fails. Same for `update()`.

roycaihw replied:
nvm, was thinking of `self.configmap_reference`

roycaihw reviewed (commented):
Thanks for generalizing the leader_election_record. Overall the implementation looks good to me. Please add a test.

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -101,82 +93,76 @@ def try_acquire_or_renew(self):
         if not lock_status:
-            if ast.literal_eval(lock_response.body)['code'] != 404:
-                logging.info("Error retrieving resource lock {} as {}".format(self.election_config.lock.name, lock_response.reason))
+            if json.loads(old_election_record.body)['code'] != 404:
```
please avoid magic constant 404, instead use NOT_FOUND

yliaog commented on leaderelection/leaderelection.py:
```diff
@@ -0,0 +1,168 @@
+            logging.info("{} successfully acquired lease".format(self.election_config.lock.identity))
+
+            # Start leading and call OnStartedLeading()
+            threading.Thread(target=self.election_config.onstarted_leading, daemon=True).start()
```
please add some documentation to the class to describe the threading behavior. i.e. the thread is running until onstarted_leading returns if it does return, it's not safe to run this leader election more than once in a process, etc.

Invictus17 commented:
/remove-lifecycle stale

Invictus17 commented:
@yliaog @roycaihw In my latest push I've added the tests but the travis ci report is indicating a few failures. Any idea why some of these tests are failing?
test result: https://travis-ci.org/github/kubernetes-client/python-base/builds/751412124

yliaog commented:
https://github.com/kubernetes-client/python-base/pull/222 fixed the configmap test failure

roycaihw commented on leaderelection/leaderelection_test.py:
```diff
@@ -0,0 +1,247 @@
+                                         onstopped_leading=on_stopped_leading_A)
+
+        # Enter leader election
+        leaderelection.LeaderElection(config_A).run()
```
Since `run()` blocks until the leader election ends, B won't start until A hits `renew_count_max`. Could you make the two clients run in parallel to make sure the leader election handles concurrency properly?

roycaihw commented on leaderelection/leaderelection_test.py:
```diff
@@ -0,0 +1,247 @@
+                             "try update record",
+                             "update record",
+                             "try update record",
+                             "try update record"])
```
I had some trouble understanding what this test (`test_Leader_election_with_renew_deadline `) does. Could you add some comments to explain how it's related to `renew_deadline` and what's the expected behavior? Thanks

Invictus17 replied:
Thanks @roycaihw, sure I'll add some comments. The expected behavior is to check if the leader stops leading if it fails to update the lock after the renew_deadline.

roycaihw commented on leaderelection/leaderelection_test.py:
```diff
@@ -0,0 +1,267 @@
+    on update:  zzz s
+    on try update:  3s
+    on update: zzz s 
+    on try update:  4.5s
```
Before this try, the timeout was set to be 4.5 + renew_deadline = 6.5s. After two failed tries (w/ sleep), the leader exited at 4.5+1.5*2=7.5s.

roycaihw commented:
LGTM. Please squash the commits into reasonable parts.

roycaihw commented on leaderelection/leaderelection_test.py:
```diff
@@ -0,0 +1,267 @@
+        self.name = name
+        self.namespace = namespace
+        self.identity = str(identity)
+        self.lock = threading.RLock()
```
is this lock per MockResourceLock? I'd expect it to be shared between all MockResourceLocks in one test

Invictus17 replied:
No, it should be shared. I'll update it. Thanks!

roycaihw commented:
/lgtm
/approve

Thanks!
